### PR TITLE
Enable downloading page sources with ?raw.

### LINKED ISSUE #1906: Feature request: Download raw Markdown source for a page
I would like an API for grabbing the raw, unrendered Markdown source for a page, like the "Raw" button on GitHub:

https://raw.githubusercontent.com/gollum/gollum/master/CONTRIBUTING.md

I understand I can access the underlying file on disk, or use the Edit button and copy the contents of the page.

This would make it easy to access a page from a script, or to export it to make a plain text attachment on an email, etc.

### PR DESCRIPTION
**Pull request by tstein**
resolves #1906

Folks on that issue sounded okay with this implementation, and imo it's better to break literally no one than to expand the set of paths that clash with user content.

### CODE REVIEW

File: lib/gollum/app.rb
Comment (bartkamphorst) [inline lib/gollum/app.rb:592]: I don't have all the details at the moment, but would it be more straightforward to check if `params.has_key?("raw")` here in this route, and, if so, call a new method `show_raw_contents(params[:splat].first)`?

Comment (tstein) [reply lib/gollum/app.rb:592]: It makes sense to me to use `show_file` because `raw` essentially means "retrieve that wiki page and skip rendering" - basically just "give me the file". If you accept that, `show_page_or_file` looks like the perfect place to use the existing call to `show_file` and also to only give `?raw` an effect when you know you're serving an actual wiki page (and not a redirect to another page or a creation view).

If you don't buy that, I'll defer to you, of course. :)

File: lib/gollum/app.rb
Comment (bartkamphorst) [inline lib/gollum/app.rb:616]: Would it be sensible to default `raw` to `false` in this method signature?

Comment (benjaminwil) [reply lib/gollum/app.rb:616]: Based on the way it's implemented, it should always be provided a value. So it's technically okay. But I think there's a benefit to providing a default value and ensuring readers see what the "happy path" (read as: non-raw) looks like.

I wouldn't block merge, but I think that this implementation could be simpler by extracting raw-specific concerns to a separate method:

```
def show_page_or_file(fullpath, raw: false)
  return show_raw_page_or_file(fullpath) if raw

  # The already-committed implementation of `#show_page_or_file`.
end

def show_raw_page_or_file(fullpath)
  # Only the parts of the if statement in this diff where the presence
  # of `raw` matters.
end
```

This results in more lines of code, but less room for error.

File: lib/gollum/app.rb
Comment (benjaminwil) [inline lib/gollum/app.rb:593]: If you decided to try out my suggested method signature, this would also need to be changed to match:

```suggestion
      show_page_or_file(params[:splat].first, raw: params.has_key?("raw"))
```

Comment (tstein) [reply lib/gollum/app.rb]: Done.

File: lib/gollum/app.rb
Comment (tstein) [reply lib/gollum/app.rb]: Added the default. Just for fun, I did your suggested version on a different branch: https://github.com/tstein/gollum/commit/02fbf7ad67e4a792a74af6aed8b6953539318219

I also did it the way you suggested - IMO the duplication necessitates some very small helpers, and overall I think it's not worth the extra volume. lmk what you prefer, it'll only take a moment to swap that commit in for this one. :)

File: lib/gollum/app.rb
Comment (tstein) [reply lib/gollum/app.rb:592]: Done, and see the other thread - I took a stab at the other approach in another branch, linked there.

File: lib/gollum/app.rb
Comment (bartkamphorst) [reply lib/gollum/app.rb]: I like the idea of extracting a `page_not_found` helper (from the alternative branch), I think that improves the readability of the logic overall. More generally, I'm still on board with @tstein 's logic here and using `show_file` to treat a page as a file when `raw` is true.

Comment (tstein) [reply lib/gollum/app.rb]: Works for me. This PR is now for https://github.com/tstein/gollum/commit/02fbf7ad67e4a792a74af6aed8b6953539318219.

**Comment by dometto** [comment]
Thanks for working on this @tstein! Could I ask you to add a small line to `LATEST_CHANGES.md` highlighting your contribution? We should update the PR template to request people to do so. :)

**Comment by tstein** [comment]
Sure thing - done.

**Review by bartkamphorst — APPROVED** [review]

**Comment by bartkamphorst** [comment]
Okay to merge, @benjaminwil ?

**Comment by benjaminwil** [comment]
Yes! Thanks @tstein

**Comment by bartkamphorst** [comment]
Thanks @tstein , we appreciate your contribution, as well as your patience and willingness to take a slightly modified approach. If there are any other issues you'd like to take a stab at, please let us know and we'd be happy to discuss options and help you on your way.

**Comment by tstein** [comment]
:+1: Thanks for the review and for your work on gollum, I'm a big fan. :)

**Pull request merged**
